# Practical Application III: Comparing Classifiers

**Overview**: In this practical application, your goal is to compare the performance of the classifiers we encountered in this section, namely K Nearest Neighbor, Logistic Regression, Decision Trees, and Support Vector Machines.  We will utilize a dataset related to marketing bank products over the telephone.  



### Getting Started

Our dataset comes from the UCI Machine Learning repository [link](https://archive.ics.uci.edu/ml/datasets/bank+marketing).  The data is from a Portugese banking institution and is a collection of the results of multiple marketing campaigns.  We will make use of the article accompanying the dataset [here](CRISP-DM-BANK.pdf) for more information on the data and features.



### Problem 1: Understanding the Data

To gain a better understanding of the data, please read the information provided in the UCI link above, and examine the **Materials and Methods** section of the paper.  How many marketing campaigns does this data represent?

The dataset represents 17 direct marketing campaigns conducted by a Portuguese banking institution between May 2008 and November 2010. The campaigns involved telephone contacts with clients to determine if they would subscribe to a term deposit product.

### Problem 2: Read in the Data

Use pandas to read in the dataset `bank-additional-full.csv` and assign to a meaningful variable name.

In [1]:
import pandas as pd

In [2]:
import os
os.getcwd()

'/Users/rahimmerchant/Downloads/Assignment_17'

In [3]:
df = pd.read_csv('bank-additional-full.csv', sep=';')

In [4]:
df.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


In [5]:
df.shape

(41188, 21)

### Problem 3: Understanding the Features


Examine the data description below, and determine if any of the features are missing values or need to be coerced to a different data type.


```
Input variables:
# bank client data:
1 - age (numeric)
2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
5 - default: has credit in default? (categorical: 'no','yes','unknown')
6 - housing: has housing loan? (categorical: 'no','yes','unknown')
7 - loan: has personal loan? (categorical: 'no','yes','unknown')
# related with the last contact of the current campaign:
8 - contact: contact communication type (categorical: 'cellular','telephone')
9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
# other attributes:
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
14 - previous: number of contacts performed before this campaign and for this client (numeric)
15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')
# social and economic context attributes
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)
17 - cons.price.idx: consumer price index - monthly indicator (numeric)
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)
20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')
```



In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

After analyzing the dataset description and the output of df.info(), the data types appear to be appropriate. Numerical variables are stored as numeric data types (int64 and float64), while categorical variables are stored as object types.

The dataset does not contain missing values (NaN). However, several categorical features contain the value "unknown", which represents missing or unavailable information.

### Problem 4: Understanding the Task

After examining the description and data, your goal now is to clearly state the *Business Objective* of the task.  State the objective below.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

The goal of this analysis is to predict whether a customer will subscribe to a term deposit based on information about the customer, previous marketing interactions, and economic conditions at the time of contact. By identifying customers who are more likely to subscribe, the bank can focus its marketing efforts more effectively, reduce time spent contacting low probability customers, and improve the overall success of future campaigns. To determine the most effective approach, several classification models will be developed and compared, including K-Nearest Neighbours, Logistic Regression, Decision Trees, and Support Vector Machines.

### Problem 5: Engineering Features

Now that you understand your business objective, we will build a basic model to get started.  Before we can do this, we must work to encode the data.  Using just the bank information features, prepare the features and target column for modeling with appropriate encoding and transformations.

In [8]:
X = df[['age',
        'job',
        'marital',
        'education',
        'default',
        'housing',
        'loan']]

y = df['y']

In [9]:
X.head()

,age,job,marital,education,default,housing,loan
0,56,housemaid,married,basic.4y,no,no,no
1,57,services,married,high.school,unknown,no,no
2,37,services,married,high.school,no,yes,no
3,40,admin.,married,basic.6y,no,no,no
4,56,services,married,high.school,no,no,yes


In [10]:
y = y.map({
    'no': 0,
    'yes': 1
})

In [11]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: y, dtype: int64

In [12]:
X = pd.get_dummies(
    X,
    drop_first=True
)

In [13]:
X.head()

,age,job_blue-collar,job_entrepreneur,job_housemaid,job_management,job_retired,job_self-employed,job_services,job_student,job_technician,...,education_illiterate,education_professional.course,education_university.degree,education_unknown,default_unknown,default_yes,housing_unknown,housing_yes,loan_unknown,loan_yes
0,56,False,False,True,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,57,False,False,False,False,False,False,True,False,False,...,False,False,False,False,True,False,False,False,False,False
2,37,False,False,False,False,False,False,True,False,False,...,False,False,False,False,False,False,False,True,False,False
3,40,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
4,56,False,False,False,False,False,False,True,False,False,...,False,False,False,False,False,False,False,False,False,True


In [14]:
X.shape

(41188, 28)

For this analysis, the customer information features were used as the predictor variables and the subscription outcome (y) was used as the target variable. The target was converted from "yes" and "no" to a binary format (1 and 0). Because several of the features were categorical, pd.get_dummies() was applied to transform them into numerical variables that could be used by the classification models.

### Problem 6: Train/Test Split

With your data prepared, split it into a train and test set.

In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [16]:
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (32950, 28)
X_test shape: (8238, 28)
y_train shape: (32950,)
y_test shape: (8238,)


### Problem 7: A Baseline Model

Before we build our first model, we want to establish a baseline.  What is the baseline performance that our classifier should aim to beat?

In [17]:
y.value_counts(normalize=True)

y
0    0.887346
1    0.112654
Name: proportion, dtype: float64

In [18]:
baseline = y.value_counts(normalize=True).max()

print(f"Baseline Accuracy: {baseline:.3f}")

Baseline Accuracy: 0.887


The baseline accuracy represents the performance we would achieve by always predicting the most common outcome. Since the majority of customers in this dataset did not subscribe to a term deposit, predicting "no" for every customer results in an accuracy of about 88.7%. Any model we build should be able to outperform this baseline to show that it is capturing useful information from the data.

### Problem 8: A Simple Model

Use Logistic Regression to build a basic model on your data.  

In [19]:
from sklearn.linear_model import LogisticRegression

In [20]:
log_model = LogisticRegression(max_iter=1000)

log_model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [21]:
y_pred = log_model.predict(X_test)

In [22]:
from sklearn.metrics import accuracy_score

log_accuracy = accuracy_score(y_test, y_pred)

print(f"Logistic Regression Accuracy: {log_accuracy:.4f}")

Logistic Regression Accuracy: 0.8874


Logistic Regression model was trained using the training data and evaluated on the test set. The model had an accuracy of 88.74%, which was compared against the baseline accuracy of approximately 88.7%. Since the model was identical to the baseline, it suggests that the model is not providing significant improvement just by predicting majority class.

### Problem 9: Score the Model

What is the accuracy of your model?

The Logistic Regression model achieved an accuracy of 88.74%.

### Problem 10: Model Comparisons

Now, we aim to compare the performance of the Logistic Regression model to our KNN algorithm, Decision Tree, and SVM models.  Using the default settings for each of the models, fit and score each.  Also, be sure to compare the fit time of each of the models.  Present your findings in a `DataFrame` similar to that below:

| Model | Train Time | Train Accuracy | Test Accuracy |
| ----- | ---------- | -------------  | -----------   |
|     |    |.     |.     |

In [23]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC

from sklearn.metrics import accuracy_score

import time

In [24]:
start = time.time()

log_model = LogisticRegression(max_iter=1000)
log_model.fit(X_train, y_train)

log_train_time = time.time() - start

log_train_acc = log_model.score(X_train, y_train)
log_test_acc = log_model.score(X_test, y_test)

In [25]:
start = time.time()

knn_model = KNeighborsClassifier()
knn_model.fit(X_train, y_train)

knn_train_time = time.time() - start

knn_train_acc = knn_model.score(X_train, y_train)
knn_test_acc = knn_model.score(X_test, y_test)

In [26]:
start = time.time()

dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)

dt_train_time = time.time() - start

dt_train_acc = dt_model.score(X_train, y_train)
dt_test_acc = dt_model.score(X_test, y_test)

In [27]:
start = time.time()

svm_model = SVC(random_state=42)

svm_model.fit(X_train, y_train)

svm_train_time = time.time() - start

svm_train_acc = svm_model.score(X_train, y_train)
svm_test_acc = svm_model.score(X_test, y_test)

In [28]:
results = pd.DataFrame({
    'Model': [
        'Logistic Regression',
        'KNN',
        'Decision Tree',
        'SVM'
    ],
    'Train Time': [
        log_train_time,
        knn_train_time,
        dt_train_time,
        svm_train_time
    ],
    'Train Accuracy': [
        log_train_acc,
        knn_train_acc,
        dt_train_acc,
        svm_train_acc
    ],
    'Test Accuracy': [
        log_test_acc,
        knn_test_acc,
        dt_test_acc,
        svm_test_acc
    ]
})

results

,Model,Train Time,Train Accuracy,Test Accuracy
0,Logistic Regression,0.810732,0.887344,0.887351
1,KNN,0.019141,0.890653,0.878733
2,Decision Tree,0.194273,0.917056,0.864045
3,SVM,24.870127,0.887344,0.887351


The Logistic Regression and SVM models achieved the highest test accuracy at 88.74%, while KNN achieved 87.87% and the Decision Tree achieved 86.40%. Although the Decision Tree had the highest training accuracy at 91.71%, its lower test accuracy suggests that it may be overfitting the training data.

KNN had the shortest running time, followed by the Decision Tree and Logistic Regression. The SVM model required significantly more running time.

Overall, Logistic Regression provided the best balance between model performance and training time. While SVM matched its accuracy, it required substantially more running time and therefore did not provide a meaningful performance advantage in this case.

### Problem 11: Improving the Model

Now that we have some basic models on the board, we want to try to improve these.  Below, we list a few things to explore in this pursuit.


- Hyperparameter tuning and grid search.  All of our models have additional hyperparameters to tune and explore.  For example the number of neighbors in KNN or the maximum depth of a Decision Tree.  
- Adjust your performance metric

In [29]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import balanced_accuracy_score

In [30]:
knn_params = {
    'n_neighbors': [3, 5, 7, 9, 11, 15]
}

knn_grid = GridSearchCV(
    KNeighborsClassifier(),
    knn_params,
    cv=5,
    scoring='balanced_accuracy'
)

knn_grid.fit(X_train, y_train)

knn_best = knn_grid.best_estimator_

knn_best_train = balanced_accuracy_score(y_train, knn_best.predict(X_train))
knn_best_test = balanced_accuracy_score(y_test, knn_best.predict(X_test))

print("Best KNN Parameters:", knn_grid.best_params_)
print("KNN Balanced Train Accuracy:", knn_best_train)
print("KNN Balanced Test Accuracy:", knn_best_test)

Best KNN Parameters: {'n_neighbors': 3}
KNN Balanced Train Accuracy: 0.6056128741145793
KNN Balanced Test Accuracy: 0.5362334308222085


In [31]:
dt_params = {
    'max_depth': [3, 5, 7, 10, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 5, 10]
}

dt_grid = GridSearchCV(
    DecisionTreeClassifier(random_state=42),
    dt_params,
    cv=5,
    scoring='balanced_accuracy'
)

dt_grid.fit(X_train, y_train)

dt_best = dt_grid.best_estimator_

dt_best_train = balanced_accuracy_score(y_train, dt_best.predict(X_train))
dt_best_test = balanced_accuracy_score(y_test, dt_best.predict(X_test))
print("Best Decision Tree Parameters:", dt_grid.best_params_)
print("Decision Tree Balanced Train Accuracy:", dt_best_train)
print("Decision Tree Balanced Test Accuracy:", dt_best_test)

Best Decision Tree Parameters: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 5}
Decision Tree Balanced Train Accuracy: 0.6047680499190945
Decision Tree Balanced Test Accuracy: 0.5218120548139064


In [32]:
log_params = {
    'C': [0.01, 0.1, 1, 10, 100]
}

log_grid = GridSearchCV(
    LogisticRegression(max_iter=1000),
    log_params,
    cv=5,
    scoring='balanced_accuracy'
)

log_grid.fit(X_train, y_train)

log_best = log_grid.best_estimator_

log_best_train = balanced_accuracy_score(y_train, log_best.predict(X_train))
log_best_test = balanced_accuracy_score(y_test, log_best.predict(X_test))

print("Best Logistic Regression Parameters:", log_grid.best_params_)
print("Logistic Regression Balanced Train Accuracy:", log_best_train)
print("Logistic Regression Balanced Test Accuracy:", log_best_test)

Best Logistic Regression Parameters: {'C': 0.01}
Logistic Regression Balanced Train Accuracy: 0.5
Logistic Regression Balanced Test Accuracy: 0.5


In [33]:
svm_params = {
    'C': [0.1, 1, 10]
}

svm_grid = GridSearchCV(
    SVC(),
    svm_params,
    cv=3,
    scoring='balanced_accuracy'
)

svm_grid.fit(X_train, y_train)

svm_best = svm_grid.best_estimator_

svm_best_train = balanced_accuracy_score(y_train, svm_best.predict(X_train))
svm_best_test = balanced_accuracy_score(y_test, svm_best.predict(X_test))

print("Best SVM Parameters:", svm_grid.best_params_)
print("SVM Balanced Train Accuracy:", svm_best_train)
print("SVM Balanced Test Accuracy:", svm_best_test)

Best SVM Parameters: {'C': 10}
SVM Balanced Train Accuracy: 0.5001175972429597
SVM Balanced Test Accuracy: 0.5018815746025757


In [34]:
improved_results = pd.DataFrame({
    'Model': [
        'Logistic Regression',
        'KNN',
        'Decision Tree',
        'SVM'
    ],
    'Best Parameters': [
        log_grid.best_params_,
        knn_grid.best_params_,
        dt_grid.best_params_,
        svm_grid.best_params_
    ],
    'Balanced Train Accuracy': [
        log_best_train,
        knn_best_train,
        dt_best_train,
        svm_best_train
    ],
    'Balanced Test Accuracy': [
        log_best_test,
        knn_best_test,
        dt_best_test,
        svm_best_test
    ]
})

improved_results

,Model,Best Parameters,Balanced Train Accuracy,Balanced Test Accuracy
0,Logistic Regression,{'C': 0.01},0.500000,0.500000
1,KNN,{'n_neighbors': 3},0.605613,0.536233
2,Decision Tree,"{'max_depth': None, 'min_samples_leaf': 1, 'mi...",0.604768,0.521812
3,SVM,{'C': 10},0.500118,0.501882


After tuning the models and evaluating them with balanced accuracy, the results showed that KNN performed the best with a balanced test accuracy of 53.6%. Decision Tree followed at 52.2%, while Logistic Regression and SVM both performed close to 50%.

These results suggest that the customer information features alone do not provide a strong signal for predicting whether a customer will subscribe to a term deposit. While KNN showed some improvement over the other models, overall performance by the models remained relatively low. This indicates that additional features may be needed to improve predictive performance.

##### Questions